File: apps/web/lib/patients.ts
import { apiFetch } from "@/lib/api";

export interface PatientListItem {
  id: string;
  first_name: string;
  last_name: string;
  mrn: string | null;
  city: string | null;
  state: string | null;
  created_at: string;
}

export interface PatientListResponse {
  items: PatientListItem[];
  total: number;
  limit: number;
  offset: number;
}

export interface PatientDetail {
  id: string;
  first_name: string;
  last_name: string;
  dob: string;
  phone: string;
  email: string | null;
  address_line1: string | null;
  address_line2: string | null;
  city: string | null;
  state: string | null;
  zip_code: string | null;
  mrn: string | null;
  created_at: string;
}

export async function listPatients(q?: string): Promise<PatientListResponse> {
  const qs = q?.trim() ? `?q=${encodeURIComponent(q.trim())}` : "";
  const resp = await apiFetch(`/api/patients/${qs}`);
  if (!resp.ok) return { items: [], total: 0, limit: 0, offset: 0 };
  return (await resp.json()) as PatientListResponse;
}

export async function getPatient(id: string): Promise<PatientDetail | null> {
  const resp = await apiFetch(`/api/patients/${id}`);
  if (resp.status === 404) return null;
  if (!resp.ok) throw new Error(`patient detail fetch failed: ${resp.status}`);
  return (await resp.json()) as PatientDetail;
}

File: apps/api/app/schemas/patients.py
"""Pydantic schemas for the patient registry."""

from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, ConfigDict


class PatientListItem(BaseModel):
    """Row in the patient registry list. Deliberately minimal PHI — no
    dob/phone in bulk responses; those surface on the detail view."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    first_name: str
    last_name: str
    mrn: str | None
    city: str | None
    state: str | None
    created_at: datetime


class PatientListResponse(BaseModel):
    items: list[PatientListItem]
    total: int
    limit: int
    offset: int


class PatientDetail(BaseModel):
    """Full demographics for a single patient. dob/phone decrypt via the
    EncryptedString TypeDecorator when read from the ORM."""

    model_config = ConfigDict(from_attributes=True)

    id: UUID
    first_name: str
    last_name: str
    dob: str
    phone: str
    email: str | None
    address_line1: str | None
    address_line2: str | None
    city: str | None
    state: str | None
    zip_code: str | None
    mrn: str | None
    created_at: datetime
